src/Bottle/provider.js

```diff
@@ -93,6 +94,25 @@ var createProvider = function createProvider(name, Provider) {
     return this;
 };
 
+var removeProviderMap = function resetProvider(name) {
+    this.providerMap[name] = false;
+    delete this.container[name];
+    delete this.container[name + 'Provider'];
+};
+
+var resetProviders = function resetProviders() {
+    var providers = this.originalProviders;
+    Object.keys(providers).forEach(function(provider) {
+        var parts = provider.split('.');
+        if (parts.length > 1) {
+            removeProviderMap.call(this, parts[0]);
+            parts.forEach(removeProviderMap.bind(getNestedBottle.call(this, parts[0])));
+        }
+        removeProviderMap.call(this, provider);
```

Comment:
@young-steveo It is necessary for sub-containers. For example, a provider for `Thing.Something` creates: `Thing`, `Something`, and `Thing.Something` into the container providers, which all must be removed. In addition, the sub-containers are created on a different bottle instance, hence the `getNetstedBottle` binding.